### PR TITLE
Fix PHP Bug #78950: day-of-year (z) doesn't handle leap years

### PR DESCRIPTION
**Pull request by IMSoP**
Original report: https://bugs.php.net/bug.php?id=79580

This patch changes behaviour in two ways:

* First, it delays applying day-of-year until the end of parsing,
  so that it is normalised with respect to a year specified later.
* Second, it makes mixing day-of-year with month or day-of-month
  an error, rather than silently changing behaviour in that case.

I'm not sure how to add tests for this here, but the existing tests all pass, and a branch of PHP with passing PHPT tests is here: https://github.com/php/php-src/compare/PHP-7.3...IMSoP:bug-79580

### CODE REVIEW

File: parse_date.re [hunk 1]
```diff
@@ -2121,12 +2122,8 @@ timelib_time *timelib_parse_from_format_with_map(char *format, char *string, siz
 				break;
 			case TIMELIB_FORMAT_DAY_OF_YEAR: /* day of year - resets month (0 based) - also initializes everything else to !TIMELIB_UNSET */
 				TIMELIB_CHECK_NUMBER;
-				if ((tmp = timelib_get_nr((char **) &ptr, 3)) == TIMELIB_UNSET) {
+				if ((day_of_year = timelib_get_nr((char **) &ptr, 3)) == TIMELIB_UNSET) {
```
Comment (derickr) [inline parse_date.re:2125]: Please separate out the assignment and the equivalence test into two lines.

File: parse_date.re [hunk 2]
```diff
@@ -2400,6 +2397,19 @@ timelib_time *timelib_parse_from_format_with_map(char *format, char *string, siz
 		}
 	}
 
+	/* Handle day of year */
+	if ( day_of_year != TIMELIB_UNSET ) {
```
Comment (derickr) [inline parse_date.re:2401]: No extra spaces inside `(` and `)`, please.

File: parse_date.re [hunk 3]
```diff
@@ -2400,6 +2397,19 @@ timelib_time *timelib_parse_from_format_with_map(char *format, char *string, siz
...
+		/* Check for mixing day-of-year with month or day-of-month */
+		if (s->time->m != TIMELIB_UNSET || s->time->d != TIMELIB_UNSET) {
+			add_pbf_error(s, TIMELIB_ERR_MIX_DAY_OF_YEAR_WITH_M_D, "Mixing of day-of-year with month or day is not allowed", string, ptr);
+		}
+		else {
```
Comment (derickr) [inline parse_date.re:2406]: `} else {` goes on one line. I would also prefer that the part in the `if` contains the success case, and the part in the `else` the failure case.

**Comment by cmb69** [comment]
https://bugs.php.net/bug.php?id=79580 has been closed in the meantime; is this PR still relevant?

**Comment by derickr** [comment]
It is no longer relevant, the fix has been introduced in PHP 8.1 already, where we started throwing a `"A 'day of year' can only come after a year has been found"` error for this instead.

**Pull request closed without merging**